### PR TITLE
Move views into separate files for easier modification

### PR DESCRIPTION
**Pull request by DaanHoogland**
### Description

we moved to a new way of maintaining views (see #7417)
not all views were added in that PR so this pr is intended to do so

- [x] account_netstats_view
- [x] account_view
- [x] account_vmstats_view
- [x] affinity_group_view
- [x] domain_view
- [x] event_view
- [x] free_ip_view
- [x] image_store_view
- [x] instance_group_view
- [x] last_annotation_view
- [x] mshost_view
- [x] project_account_view
- [x] project_invitation_view
- [x] project_view
- [x] resource_tag_view
- [x] security_group_view
- [x] volume_view
- [x] vpc_offering_view

##### EXPECTED RESULTS

~~~
$ ls -1 engine/schema/src/main/resources/META-INF/db/views/
cloud.account_netstats_view.sql
cloud.account_view.sql
cloud.account_vmstats_view.sql
cloud.affinity_group_view.sql
cloud.async_job_view.sql
cloud.data_center_view.sql
cloud.disk_offering_view.sql
cloud.domain_router_view.sql
cloud.domain_view.sql
cloud.event_view.sql
cloud.free_ip_view.sql
cloud.host_view.sql
cloud.image_store_view.sql
cloud.instance_group_view.sql
cloud.last_annotation_view.sql
cloud.mshost_view.sql
cloud.network_offering_view.sql
cloud.project_account_view.sql
cloud.project_invitation_view.sql
cloud.project_view.sql
cloud.resource_tag_view.sql
cloud.security_group_view.sql
cloud.service_offering_view.sql
cloud.snapshot_view.sql
cloud.snapshot_view.sql
cloud.storage_pool_view.sql
cloud.storage_pool_view.sql
cloud.template_view.sql
cloud.template_view.sql
cloud.user_view.sql
cloud.user_view.sql
cloud.user_vm_view.sql
cloud.user_vm_view.sql
cloud.volume_view.sql
cloud.vpc_offering_view.sql
~~~

##### ACTUAL RESULTS

~~~
$ ls -1 engine/schema/src/main/resources/META-INF/db/views/
cloud.async_job_view.sql
cloud.data_center_view.sql
cloud.disk_offering_view.sql
cloud.domain_router_view.sql
cloud.host_view.sql
cloud.network_offering_view.sql
cloud.service_offering_view.sql
cloud.snapshot_view.sql
cloud.storage_pool_view.sql
cloud.template_view.sql
cloud.user_view.sql
cloud.user_vm_view.sql
~~~

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [X] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@GutoVeronezi , can you advice if this is all that is needed to put views under the new maintenance regime? I really only want volume_view for now, so is there any reason not to do them all right now?
thanks

File: engine/schema/src/main/resources/META-INF/db/views/cloud.volume_view.sql [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    (((`resource_tags`.`resource_id` = `volumes`.`id`)
+        and (`resource_tags`.`resource_type` = 'Volume'))))
+LEFT JOIN `async_job`ON
+    (((`async_job`.`instance_id` = `volumes`.`id`)
+        and (`async_job`.`instance_type` = 'Volume')
+            and (`async_job`.`job_status` = 0))))
+LEFT JOIN `account` `resource_tag_account`ON
+    ((`resource_tag_account`.`id` = `resource_tags`.`account_id`)))
+LEFT JOIN `domain` `resource_tag_domain`ON
+    ((`resource_tag_domain`.`id` = `resource_tags`.`domain_id`)));
```
Comment (DaanHoogland) [inline engine/schema/src/main/resources/META-INF/db/views/cloud.volume_view.sql:157]:
```suggestion
    ((`resource_tag_domain`.`id` = `resource_tags`.`domain_id`)));
```

**Comment by GutoVeronezi** [comment]
> @GutoVeronezi , can you advice if this is all that is needed to put views under the new maintenance regime? I really only want volume_view for now, so is there any reason not to do them all right now? thanks

By the design we created, creating the view's file under the `views` directory should be enough.

PR #7417's main purpose was to introduce the concept. The idea was to create the new files as we would change the views; however, there is no problem in creating them right away.

**Comment by DaanHoogland** [comment]
> PR #7417's main purpose was to introduce the concept.

I know, or better I remember now ;)
> The idea was to create the new files as we would change the views; however, there is no problem in creating them right away.

In hindsight it might be a bit confusing and lead to review errors . I'll get myself in an agry streak soon and add the rest.

**Comment by GutoVeronezi** [comment]
> In hindsight it might be a bit confusing and lead to review errors . 

Indeed

**Comment by weizhouapache** [comment]
> > PR #7417's main purpose was to introduce the concept.
> 
> I know, or better I remember now ;)
> 
> > The idea was to create the new files as we would change the views; however, there is no problem in creating them right away.
> 
> In hindsight it might be a bit confusing and lead to review errors . I'll get myself in an agry streak soon and add the rest.

@DaanHoogland @GutoVeronezi 

IMHO, ideally the database upgrade process should be
- drop procedures if exist, and create procedures
- perform database upgrade scripts (in .sql files) and data migration (in .java files), which should not depend on views
- create views

#7417 has introduced the concept to create views as last step, which is good.
however, we should also consider 
- recreating all views might take quite long time if the database if huge
- the upgrade script or data migration might rely on some views (need double-check)

**Comment by DaanHoogland** [comment]
> * recreating all views might take quite long time if the database if huge

so we should add intelligence as to which views have changed
> * the upgrade script or data migration might rely on some views (need double-check)

well, this should not be a problem as all views are still there while other upgrades occur. I think it is the programmers responsibility to have their dependencies straight.

the first one is a valid consideration but out of scope here I think.

**Comment by GutoVeronezi** [comment]
> * recreating all views might take quite long time if the database if huge

View structures should not be affected by the database size. They are structures that represent queries; thus, not relying on the amount of data.

> * the upgrade script or data migration might rely on some views (need double-check)

Regarding this one, you can check my comment on PR https://github.com/apache/cloudstack/pull/7417#issuecomment-1741821866,  which answers the same question you made.

**Comment by weizhouapache** [comment]
> > * recreating all views might take quite long time if the database if huge
> 
> View structures should not be affected by the database size. They are structures that represent queries; thus, not relying on the amount of data.
> 
good to know it.

> > * the upgrade script or data migration might rely on some views (need double-check)
> 
> Regarding this one, you can check my comment on PR https://github.com/apache/cloudstack/pull/7417#issuecomment-1741821866,  which answers the same question you made.
> 
> 
Yes, I remember it.
I am ok with moving all views to separated files.  We need to test the upgrade of different cloudstack versions.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test alma9 kvm-alma9

**Review by rohityadavcloud — APPROVED** [review]
Refactoring makes sense

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Pull request merged**